charmcraft/parts.py

```diff
@@ -0,0 +1,209 @@
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        venv_dir = self._part_info.part_install_dir / self._part_info.venv_dir
+        return {
+            "PATH": "{}/bin:${{PATH}}".format(str(venv_dir)),
+        }
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        venv_dir = self._part_info.part_install_dir / self._part_info.venv_dir
+        pip_install_cmd = f"pip install --target={venv_dir}"
+        options = cast(CharmPluginProperties, self._options)
+        commands = []
+
+        if not options.charm_allow_pip_binary:
+            pip_install_cmd += " --no-binary :all:"
+
+        if options.charm_python_packages:
+            python_packages = " ".join(
+                [shlex.quote(pkg) for pkg in options.charm_python_packages]
+            )
+            python_packages_cmd = f"{pip_install_cmd} {python_packages}"
+            commands.append(python_packages_cmd)
+
+        if options.charm_requirements:
+            requirements = " ".join(f"-r {r!r}" for r in options.charm_requirements)
+            requirements_cmd = f"{pip_install_cmd} {requirements}"
+            commands.append(requirements_cmd)
+
+        install_dir = self._part_info.part_install_dir
+        commands.append(
+            'cp --archive --link --no-dereference . "{}"'.format(install_dir)
+        )
+
+        return commands
+
+
+def setup_parts():
+    """Initialize craft-parts plugins."""
+    plugins.register({"charm": CharmPlugin})
+
+
+def validate_part(data: Dict[str, Any]) -> None:
+    """Validate the given part data against common and plugin models.
+
+    :param data: The part data to validate.
+    """
+    if not isinstance(data, dict):
+        raise TypeError("value must be a dictionary")
+
+    spec = data.copy()
+    plugin_name = spec.get("plugin", "")
+    if not plugin_name:
+        raise ValueError("'plugin' not defined")
+
+    plugin_class = plugins.get_plugin_class(plugin_name)
+
+    # validate plugin properties
+    plugin_class.properties_class.unmarshal(spec)
+
+    # validate common part properties
+    plugins.strip_plugin_properties(spec, plugin_name=plugin_name)
+    PartSpec(**spec)
+
+
+class PartsLifecycle:
+    """Create and manage the parts lifecycle."""
+
+    def __init__(
+        self,
+        all_parts: Dict[str, Any],
+        *,
+        work_dir: pathlib.Path,
+        venv_dir: pathlib.Path,
+    ):
+        # set the cache dir for parts package management
+        cache_dir = BaseDirectory.save_cache_path("charmcraft")
+
+        try:
+            self._lcm = LifecycleManager(
+                {"parts": all_parts},
+                application_name="charmcraft",
+                work_dir=work_dir,
+                cache_dir=cache_dir,
+                venv_dir=venv_dir,
+            )
+            self._lcm.refresh_packages_list()
```

Comment:
Might be an optimization we can do in the future, if using a provided environment we will have a refresh apt cache.